### PR TITLE
[Local Testing] SNOW-976759 Fix implementation of mock_avg to compute the correct return type and precision

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   N/A

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Previously `mock_avg` always return a column of floating point values rounded to 5 digits after the floating point, where the result type is incorrectly set to the type of the incoming column. As a result, average of Integer column will have a return type of Integer. Average of Decimal columns will not have the correct precision and scale. This PR fixes these issues.

### CODE REVIEW

File: src/snowflake/snowpark/mock/functions.py [hunk 1]
```diff
@@ -141,22 +143,34 @@ def mock_sum(column: ColumnEmulator) -> ColumnEmulator:
...
-        if not all_item_is_none
-        else ColumnEmulator(data=[None])
-    )
-    ret.sf_type = column.sf_type
-    return ret
+    if isinstance(column.sf_type.datatype, NullType) or column.isna().all():
+        return ColumnEmulator(data=[None], sf_type=ColumnType(NullType(), True))
+    elif isinstance(column.sf_type.datatype, _IntegralType):
+        res_type = DecimalType(38, 6)
+    elif isinstance(column.sf_type.datatype, DecimalType):
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/mock/functions.py:155]: We have other "fixed" types like IntegerType, LongType and DecimalType.
Pls refer to `def calculate_type(c1: ColumnType, c2: Optional[ColumnType], op: Union[str]):`

File: src/snowflake/snowpark/mock/functions.py [hunk 2]
```diff
@@ -141,22 +143,34 @@ def mock_sum(column: ColumnEmulator) -> ColumnEmulator:
...
+            column.sf_type.datatype.precision,
+            column.sf_type.datatype.scale,
+        )
+        precision = max(38, column.sf_type.datatype.precision + 12)
+        if scale <= 6:
+            scale = scale + 6
+        elif scale < 12:
+            scale = 12
+        res_type = DecimalType(precision, scale)
+    else:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/mock/functions.py:166]: How about an explicit checking for `FloatType` and `DoubleType`?

File: src/snowflake/snowpark/mock/functions.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/functions.py:155]: Shouldn't IntegerType, LongType be handled by the case on line 153?

File: src/snowflake/snowpark/mock/functions.py [hunk 3]
```diff
@@ -141,22 +143,34 @@ def mock_sum(column: ColumnEmulator) -> ColumnEmulator:
...
+        res_type = DecimalType(38, 6)
+    elif isinstance(column.sf_type.datatype, DecimalType):
+        precision, scale = (
+            column.sf_type.datatype.precision,
+            column.sf_type.datatype.scale,
+        )
+        precision = max(38, column.sf_type.datatype.precision + 12)
+        if scale <= 6:
+            scale = scale + 6
+        elif scale < 12:
```
[the remark this answers is not in the document]
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/functions.py:163]: Then we use the original scale from input, i.e. it is unchanged from the assignment on line 156.

File: src/snowflake/snowpark/mock/functions.py (hunk 2 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/functions.py:166]: That's a good idea. Do you prefer using
```
elif instance(column.sf_type.datatype, _FractionalType)
    ...
else:
   raise Error
```
or an assert statement in the else case on line 166?

**Review by sfc-gh-yixie — APPROVED** [review]

**Review by sfc-gh-aalam — APPROVED** [review]

**Pull request merged**